### PR TITLE
Add new integration testing framework

### LINKED ISSUE #2223: Create an integration test to ensure that Elastic Agent's is running correctly with fake components
We need to create a basic integration test in the Agent repository (because we are testing Agent capability) to ensure that the agent is running correctly.
To do it without testing external dependencies, this test will run Agent with:
- Fake Input 
- Fake Shipper
...

In addition, this test should be build to be reusable by other team to test their component with the Agent (ex: testing that the shipper is running as expected in the elastic-agent-shipper repository).

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

This adds a new `pkg/testing` module that provides an integration testing framework for testing the Elastic Agent.

It provides support for multiple paths for fetching a built artifact of the Elastic Agent. Local fetching allows the test to use a local built version of the Elastic Agent. Artifact fetching allows fetching the Elastic Agent from the Elastic artifacts API.

It provides a way of prepare only the components that are to be used for the test, allow the ability to keep components that are shipped with the build of Elastic Agent or to insert custom built components for the test.

It provides a state machine management system that allows multiple states to be defined and the test ensures that the state is reached before moving on to the next state. Each state can define a `Before` or `After` that gets called before and after each state respectively. Each state can define a new configuration to use when that state is next in the machine.

## Why is it important?

This framework is to be used for testing the Elastic Agent inside of its own repository as well as allowing other repositories to test there components with the Elastic Agent.

The goal is to make testing the Elastic Agent and a component as easy as possible.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~

## Related issues

- Closes #2223

## Examples

The following are examples of using the testing framework.

### Simple Test
```
package integration_test

import (
	"context"
	"path/filepath"
	"runtime"
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"github.com/elastic/elastic-agent/pkg/component"
	"github.com/elastic/elastic-agent/pkg/control/v2/client"
        atesting "github.com/elastic/elastic-agent/pkg/testing"
)

var simpleConfig = `
outputs:
  default:
    type: elasticsearch
    hosts: [localhost:9200]
inputs:
  - id: filestream
    type: filestream
    streams:
    - id: syslog
      paths: /var/log/syslog
`

func TestElasticAgent_Filebeat(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
	defer cancel()

	f, err := atesting.NewFixture(t, "8.8.0")
	require.NoError(t, err)
	err = f.Prepare(ctx, atesting.UsableComponent{Name: "filebeat"})
	require.NoError(t, err)

	err = f.Run(ctx, atesting.State{
		Configure:  simpleConfig,
		AgentState: NewClientState(client.Healthy),
	})
	require.NoError(t, err)
}
```

## Follow Ups

The following are items that need to be followed up on but are not completed by the PR. The PR is large enough already so separating the work is important.

- Add more documentation with examples
- Add ability to run the integration tests with OGC locally
- Setup CI job to run the integration tests on each PR
- Add ability to perform install based tests (this PR only does running as in `elastic-agent run`)

### CODE REVIEW

**Comment by blakerouse** [comment]
I added a Follow Ups section to the description please review that with context on what this PR covers, all those missing items will be covered in other follow up PR's.

**Comment by jlind23** [comment]
> Setup CI job to run the integration tests on each PR

NIT: Is this a CI job or a new mage target?

**Comment by blakerouse** [comment]
@jlind23 

> Add ability to run the integration tests with OGC locally

^ new mage target

> Setup CI job to run the integration tests on each PR

^ CI job

**Review by pchila — COMMENTED** [review]
LGTM but, given the size, the only way to really review it is to try to use it (it's difficult to keep context while reviewing 3k lines of new code).

There are a couple of nitpicks here and there, also I am not completely convinced that a full-blown FSM is needed, I think a few assertions would have sufficed but maybe I am not thinking of certain testing scenarios (maybe for beats ?)

File: internal/pkg/agent/cmd/run.go [hunk 1]
```diff
@@ -62,6 +62,12 @@ func newRunCommandWithArgs(_ []string, streams *cli.IOStreams) *cobra.Command {
 		Use:   "run",
 		Short: "Start the elastic-agent.",
 		Run: func(cmd *cobra.Command, _ []string) {
+			// done very early so the encrypted store is never used
+			disableEncryptedStore, _ := cmd.Flags().GetBool("disable-encrypted-store")
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/cmd/run.go:66]: This is general purpose flag that allows users to run the Elastic Agent on Mac OS without being root.

File: pkg/testing/log.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	}
+}
+
+// Watch returns the channel that will get an error when an error is identified from the log.
+func (r *logWatcher) Watch() <-chan error {
+	return r.alert
+}
+
+// Write implements the `io.Writer` interface.
+func (r *logWatcher) Write(p []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/testing/log.go:46]: This implementation is rather simple and is a simplified version that the command runtime uses to capture stdout and stderr of a spawned process. So overall I think its is not complex, and provides a nice way of catching errors in the logs of Elastic Agent which we don't want.

File: internal/pkg/agent/application/coordinator/mocks/capability.go
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/application/coordinator/mocks/capability.go:1]: `mage check` adds that automatically, not something that I am looking to changing in this PR. I hear its already large enough ;-)

**Review by pchila — APPROVED** [review]

File: pkg/control/addr_common.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	dataPath := filepath.Join(noSyms, "data")
+	if platform == "windows" {
+		return fmt.Sprintf(`\\.\pipe\elastic-agent-%x`, sha256.Sum256([]byte(dataPath))), nil
+	}
+	socketPath := filepath.Join(dataPath, "tmp", "elastic-agent-control")
+	socketPath = fmt.Sprintf("unix://%s.sock", socketPath)
+	// unix socket path must be less than 104 characters
+	if len(socketPath) < 104 {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/control/addr_common.go:35]: The upgrade watcher uses the socket path, and we need each previous version to know where the path to the socket of the next version is present. So it is an issue to move its path.

File: changelog/fragments/1680014553-Add-an-integration-testing-framework-for-component-developers.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: feature
+
+# Change summary; a 80ish characters long description of the change.
+summary: Add an integration testing framework for component developers.
```
Comment (michalpristas) [inline changelog/fragments/1680014553-Add-an-integration-testing-framework-for-component-developers.yaml:14]: is this really something user needs to be informed about?

Comment (pierrehilbert) [reply changelog/fragments/1680014553-Add-an-integration-testing-framework-for-component-developers.yaml:14]: No it's not, I would remove the fragment.

**Review by michalpristas — APPROVED** [review]
code looks good, i managed to run fake test without an issue

**Comment by pazone** [comment]
> Add ability to run the integration tests with OGC locally

> Setup CI job to run the integration tests on each PR

Coincidentally E2E tests have the same follow ups. I couldn't find any issues related to those items.  
I'd like to help here :) 
@jlind23 @pierrehilbert

**Pull request merged**